parsing/longident.mli

```diff
@@ -28,7 +28,10 @@ type t =
 val flatten: t -> string list
 val unflatten: string list -> t option
 val last: t -> string
-val parse: string -> t
+val parse_flat: string -> t
+(** This function may return syntactically invalid {!t}.
+    If you need precise parsing function, consider using one of the functions
+    in {!Parse.Longident}  *)
```

Comment:
Sorry for asking only now, but do we actually need this function at all anymore? If we agree that changing the name (and thus breaking experimental-api-user code) is the the right choice, why don't we just get rid of it completely?

Reply:
The recommend replacement would be something like

```ocaml
   Parse.Longident.any (Lexing.from_string input)
```

or using a more precise syntactic category.

Reply:
Another choice would be to only mark `Location.parse` deprecated, instead of removing it altogether. The documentation/deprecation comment should explain clearly why the function is broken.

cc @diml, as our ocaml-migrate-parsetree spokeperson: is something terribly bad going to happen if we remove `Longident.parse` in favor of actually-correct parser entry points?

Reply:
There are a few consequences:

- if you remove the function entirely, that's going to cause quite a bit of churn in omp and a few other projects. Longident is currently not snapshotted by omp
- if you replace it by calling the lexer and parser, that's going it to make it slower and allocate more. This function tend to be called a lot in ppx rewriters. For instance, in ppxlib we have a function `evar : loc:Location.t -> string -> expression` that creates a variable expression from a string. This function calls `Longident.parse` and is used quite a bit. I have seen similar things in other projects. This could have an impact on the performance of ppx rewriters, or could be lost in the noise. It's difficult to say without testing

Personally, I'd just keep the current `Longident.parse` function for now. In the upcoming `ppx` library, we could stop relying on `Longident.parse` and then we could delete `Longident.parse` in a couple of years.

Reply:
Keeping `parse` sounds fine in this case (wouldn't it be nice if we had some kind of alerts telling users that a function is expected to be removed?). For ppxlib, wouldn't it work to expose an evar function with type: `loc;Location.t -> string list -> expression`? This function would avoid half of the work in `Longident.parse`, cover the same subclass of `Longident.t`s, and all surprising error cases.

Reply:
In fact, a `string list -> Longident.t` function already exists as `Longident.unflatten`. After thinking about the issue, I think that ppx authors should use `unflatten` instead (they can do the splitting on their side if desired, at least the semantics is clear to the code author), and I remain of the opinion that we should remove `Longident.parse`.

Reply:
(In particular, ppx authors switching to `unflatten` will have code that is backward-compatible, so transition pains should be limited.)

Reply:
I agree with you both, and the truth is maybe our `evar` function should do no parsing at all and consider that the argument is not a long identifier but just a variable name. We'll certainly think about this for the upcoming ppx project.

But right now, I'm personally not excited by the amount of downstream work this represent and I'd prefer not to have to deal with it now.

That said, we have been in this situation of few years now, so even if removing `Longident.parse` will cause some churn, it's nothing we haven't done before.

To give a shorter and clearer answer to "will something terribly bad happen if we remove Longident.parse": nothing terribly bad will happen and we would be able to cope with the change. However, given that `Longident` is not currently snapshotted by omp, this won't be straightforward. Instead, when upgrading omp we are going to need to turn our brains and think about how to handle the removal of `Longident.parse`. The solution could be quite annoying to implement.

Reply:
The present PR was revived by the fact that the use of `Longident.parse` by downstream users creates bug (in the specific example discussed it was a bug in Merlin). They rely on it, but a good fraction of those use-cases are wrong (probably the large majority of them), and it's not obvious to notice that they are wrong (it only occurs on inputs that are less frequent in the wild, eg. `( +. )` instead of `foo`). So yes removing the function would cause transition pain, but it will also force users to fix bugs that they are not aware of, so it is not just a waste a time.

Re. snapshotting Longident in ocaml-migrate-parsetree: for this particular change, we would just be removing one function. Would it be so bad if the module remained non-snapshotted? (non-snapshot?) Without snapshotting, the issue is that (probably wrong) code users write will fail to compile on recent OCaml versions, and they have to use a different function which exists on all supported version; that doesn't sound so bad to me (it's not as good as the usual m-o-p guarantees, but those only covers snapshotted modules indeed).

Reply:
IIUC, right now on invalid input `Longident.parse` doesn't fail and instead produces an invalid `Longident.t`? If yes, what about making it fail on invalid input? The error doesn't have to be fancy, it can just be `invalid_argument` and the caller has to use the proper parser if they want a proper error. That seems like the simplest solution to me.

> Re. snapshotting Longident in ocaml-migrate-parsetree: for this particular change, we would just be removing one function. Would it be so bad if the module remained non-snapshotted? (non-snapshot?) Without snapshotting, the issue is that (probably wrong) code users write will fail to compile on recent OCaml versions, and they have to use a different function which exists on all supported version; that doesn't sound so bad to me (it's not as good as the usual m-o-p guarantees, but those only covers snapshotted modules indeed).

￼That'd be fine yes, but still quite a lot of code to update.

Reply:
It is worse than that: `Longident.parse` can produce invalid outputs from valid inputs. And indeed, it never fails. I am not sure if it makes sense to raise an exception for corner cases that people *keep* forgetting.

Reply:
The result of returning an exception is that users would see preprocessing-time crashes (or merlin-query-time crashes) instead of silent bad-code generation. I think that asking extension authors to fix their code to use correct functions instead is better for the ecosystem.

Reply:
could you deprecate or put an alert on this function rather than delete it immediately?

Reply:
Deprecating the function sounds fine to me, but wouldn't it create essentially the same issue as deleting it for all dune users?

Reply:
No. Dune users will indeed get an error during development, but released packages won't start failing to build because of the deprecation. So if you have a ppx that you are not actively developing, you are free to ignore the deprecation initially.

Reply:
Ok, this sounds like an improvement. I made the change (and kept the `parse` name for now).

Reply:
Thanks!

Reply:
I'm still grumbling at the idea of keeping a function that is known to be very broken (it fails on `( +. )`, not just on extended module paths), but let's consider this a good compromise for now. I think the deprecation message should be improved. Current version:

```
    This function may parse incorrectly valid {!t} and return syntactically
    invalid {!t}s.
    If you need precise parsing function, consider using {!Parse.longident} or
    one of the specialized functions in {!Parse}.
```

The deprecation message should (1) list cases that are known to be broken, and (2) mention `unflatten` with a clear enough description of its specification that users know when to adopt it instead of calling `Parse`. In particular, ocaml-migrate-parsetree users do not have an easy access to the `Parse` module, but they can easily change to another function within `Longident`.

Proposal:

```
    This function is broken on identifiers that are not just "Word.Word.word";
    for example, it returns incorrect results on infix operators
    and extended module paths.
    
    If you know that the identifiers you parse are of this simple form
    (lists of dot-separated identifiers), you should use {!unflatten} instead
    which is safer and faster. {!unflatten} is available since OCaml 4.06.0.

    If you want to parse any identifier correctly, use the long-identifiers
    functions from the {!Parse} module, in particular {!Parse.longident}.
    They are available since OCaml 4.11, and also provide proper
    input-location support.
```

Reply:
Being more explicit on the issue at hand is clearly better.
I am not so fond of the second paragraph because the use case of `unflatten` is quite different. Maybe something like:
```
   If you want to generate long identifiers that are a list of
   dot-separated identifiers, the function {!unflatten} is safer and faster.
   {!unflatten} is available since OCaml 4.06.0.
```
in order to emphasis the contrast between `Parse.longident` (for parsing (unsafe) user inputs), and `unflatten` (for generating the most common class of longidents)?

Reply:
Yes, I like your second paragraph better.